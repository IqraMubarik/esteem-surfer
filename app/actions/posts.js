// @flow
import { Client } from 'dsteem';
import { makeGroupKeyForPosts } from '../utils/misc';
import type { postStateType, postActionType } from '../reducers/types';

export const POSTS_FETCH_BEGIN = 'POSTS_FETCH_BEGIN';
export const POSTS_FETCH_OK = 'POSTS_FETCH_OK';
export const POSTS_FETCH_ERROR = 'POSTS_FETCH_ERROR';
export const POSTS_INVALIDATE = 'POSTS_INVALIDATE';
export const POST_SET_READ = 'POST_SET_READ';
export const POST_SET_VOTED = 'POST_SET_VOTED';

const client = new Client('https://api.steemit.com');

export function fetchPosts(
  what: string,
  tag: string | null = '',
  more: boolean = false
) {
  return (
    dispatch: (action: postActionType) => void,
    getState: () => postStateType
  ) => {
    const { posts } = getState();

    const groupKey = makeGroupKeyForPosts(what, tag);

    // make sure tag is not null or undefined. it should be empty string.
    const query = {
      tag: tag || '',
      limit: 20
    };

    if (!more && posts.getIn([groupKey, 'entries']).size) {
      return;
    }

    const lastEntry = posts
      .getIn([groupKey, 'entries'])
      .valueSeq()
      .last();

    if (lastEntry) {
      query.start_author = lastEntry.author;
      query.start_permlink = lastEntry.permlink;
    }

    dispatch({
      type: POSTS_FETCH_BEGIN,
      payload: { group: groupKey }
    });

    client.database
      .getDiscussions(what, query)
      .then(resp => {
        dispatch({
          type: POSTS_FETCH_OK,
          payload: { data: resp, group: groupKey }
        });

        return resp;
      })
      .catch(e => {
        dispatch({
          type: POSTS_FETCH_ERROR,
          payload: { group: groupKey, error: e }
        });
      });
  };
}

export function invalidatePosts(what, tag = '') {
  return (dispatch: (action: postActionType) => void) => {
    const groupKey = makeGroupKeyForPosts(what, tag);

    dispatch({
      type: POSTS_INVALIDATE,
      payload: { group: groupKey }
    });
  };
}
