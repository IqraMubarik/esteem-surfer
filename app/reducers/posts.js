// @flow

import { Record, Map, OrderedMap } from 'immutable';
import { LOCATION_CHANGE } from 'react-router-redux';
import {
  POSTS_FETCH_BEGIN,
  POSTS_FETCH_OK,
  POSTS_FETCH_ERROR,
  POSTS_INVALIDATE
} from '../actions/posts';
import type { postActionType } from './types';
import filters from '../constants/filters.json';
import { makeGroupKeyForPosts } from '../utils/misc';

export const PostGroupRecord = Record({
  entries: OrderedMap({}),
  err: null,
  loading: false
});

const defaultState = Map();

export default function posts(
  state: Map = defaultState,
  action: postActionType
) {
  switch (action.type) {
    case LOCATION_CHANGE: {
      // Create post group when location changed
      const path = action.payload.pathname.split('/');
      if (path.length > 0 && filters.includes(path[1])) {
        const filter = path[1];
        const tag = path[2] || null;

        const groupKey = makeGroupKeyForPosts(filter, tag);
        if (state.get(groupKey) === undefined) {
          return state.set(groupKey, new PostGroupRecord());
        }
      }

      return state;
    }
    case POSTS_FETCH_BEGIN: {
      const groupKey = action.payload.group;

      return state
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], true);
    }
    case POSTS_FETCH_OK: {
      const groupKey = action.payload.group;
      const newEntries = action.payload.data;

      let newState = state
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], false);

      newEntries.forEach(entry => {
        if (!newState.hasIn([groupKey, 'entries', `${entry.id}`])) {
          newState = newState.setIn(
            [groupKey, 'entries', `${entry.id}`],
            entry
          );
        }
      });

      return newState;
    }
    case POSTS_FETCH_ERROR: {
      const groupKey = action.payload.group;
      return state
        .setIn([groupKey, 'err'], action.payload.error)
        .setIn([groupKey, 'loading'], false);
    }
    case POSTS_INVALIDATE: {
      const groupKey = action.payload.group;
      return state
        .setIn([groupKey, 'entries'], OrderedMap({}))
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], false);
    }
    default:
      return state;
  }
}
