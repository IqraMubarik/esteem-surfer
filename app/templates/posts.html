<nav-bar selected-cat="category" selected-tag="tag"></nav-bar>
<div class="app-container">
  <div class="app-content">
    <div class="side-tag-list">
      <header>{{ 'TAGS' | translate }}</header>
      <ul class="tag-list">
        <li ng-repeat="t in tags" ng-class="{'active': (tag===t)}"><a ng-click="tagClicked(t)">{{ t }}</a></li>
      </ul>
      <footer><a href="#">{{ _('View all tags..') }}</a></footer>
    </div>
    <div class="post-list">
      <header>
        <h1>{{ title | translate }}</h1>
      </header>
      <ul class="feed">
        <li ng-repeat="post in posts">
          <post-list-item post="post"></post-list-item>
        </li>
      </ul>
    </div>
  </div>
</div>
<app-footer></app-footer>
