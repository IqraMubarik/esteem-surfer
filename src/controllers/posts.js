export default ($scope, $routeParams, $filter, steemService, constants) => {

  // 1 get initial (cache) data
  // 2 set cache data when updated

  let category = $routeParams.category;
  let tag = $routeParams.tag;

  $scope.category = category;
  $scope.tag = tag;

  $scope.catTitle = constants.categories.find(i => i.name === category).key;
  if (tag) {
    $scope.tagTitle = $filter('capWord')(tag);
  }

  $scope.posts = [];
  let ids = [];
  let hasMore = true;

  const loadPosts = (startAuthor = null, startPermalink = null) => {
    // Convert category's fist letter to upper to use as function argument
    let by = $filter('capWord')(category);

    $scope.loadingPosts = true;
    steemService.getDiscussionsBy(by, (tag ? tag : null), startAuthor, startPermalink, constants.postListSize).then((resp) => {

      // if server returned less than 2 posts, it means end of pagination
      // comparison value is 2 because it returns at least 1 post on pagination
      if (resp.length < 2) {
        hasMore = false;
        return false;
      }

      resp.forEach((i) => {
        if (ids.indexOf(i.id) === -1) {
          $scope.posts.push(i);
        }
        ids.push(i.id);
      });

    }).catch(() => {
      // TODO: Handle catch
    }).then(() => {
      $scope.loadingPosts = false;
    });
  };

  loadPosts();

  $scope.reload = () => {
    if ($scope.loadingPosts) {
      return false;
    }
    $scope.posts = [];
    ids = [];
    loadPosts();
  };

  $scope.reachedBottom = () => {
    if ($scope.loadingPosts || !hasMore) {
      return false;
    }

    let lastPost = [...$scope.posts].pop();
    loadPosts(lastPost.author, lastPost.permlink)
  }
};
