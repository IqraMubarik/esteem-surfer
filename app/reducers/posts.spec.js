/* eslint-disable */
import { Record, Map, OrderedMap } from 'immutable';

import posts from './posts';
import { PostGroupRecord } from './posts';
import { LOCATION_CHANGE } from 'react-router-redux';

import {
  POSTS_FETCH_BEGIN,
  POSTS_FETCH_OK,
  POSTS_FETCH_ERROR,
  POSTS_INVALIDATE
} from '../actions/posts';

import deepFreeze from 'deep-freeze';

describe('posts reducer', () => {
  it('(1) should handle initial state', () => {
    expect(posts(undefined, {})).toMatchSnapshot();
  });

  it('(2) should create group key according to location. action:LOCATION_CHANGE', () => {
    const stateBefore = Map();

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(3) should create group key according to location. action:LOCATION_CHANGE', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/trending/art'
      }
    };
    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(4) should  not create group for non-filter path. action:LOCATION_CHANGE', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/lipsum/art'
      }
    };
    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(5) start fetching "trending". action:POSTS_FETCH_BEGIN', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(6) start fetching "trending". should keep old entries. action:POSTS_FETCH_BEGIN', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({ 34: { body: 'foo' }, 12: { body: 'bar' } }),
        err: null,
        loading: false
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(7) fetching "trending" completed. action:POSTS_FETCH_OK', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_OK,
      payload: {
        data: [
          {
            id: 14,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          },
          { id: 12, author: 'talhasch', permlink: 'lorem ipsum', votes: 12 },
          { id: 16, author: 'chrisbolten', permlink: 'sit amet', votes: 3 }
        ],
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(8) should not add exiting item to entries. action:POSTS_FETCH_OK', () => {
    const stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({
          '12': {
            id: 12,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          }
        }),
        err: null,
        loading: true
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_OK,
      payload: {
        data: [
          {
            id: 12,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          }
        ],
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(9) simulate load more. action:POSTS_FETCH_OK', () => {
    let stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });
    stateBefore = stateBefore.setIn(['trending', 'entries', '12'], {
      id: 12,
      author: 'talhasch',
      permlink: 'lorem ipsum',
      votes: 12
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_OK,
      payload: {
        data: [
          { id: 16, author: 'chrisbolten', permlink: 'sit amet', votes: 3 },
          { id: 3, author: 'dunsky', permlink: 'lorem', votes: 31 }
        ],
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(10) error occurred while fetching "trending". action:POSTS_FETCH_ERROR', () => {
    let stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_ERROR,
      payload: {
        group: 'trending',
        error: 'an error has occurred'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(11) invalidate "trending". action:POSTS_INVALIDATE', () => {
    let stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new PostGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_INVALIDATE,
      payload: {
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(12) should reset error. action:POSTS_INVALIDATE', () => {
    let stateBefore = Map({
      trending: new PostGroupRecord({
        entries: OrderedMap({}),
        err: 'an error',
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: POSTS_FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = posts(stateBefore, action);
    expect(res).toMatchSnapshot();
  });
});
