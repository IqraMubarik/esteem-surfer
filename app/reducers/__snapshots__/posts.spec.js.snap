// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts reducer (1) should handle initial state 1`] = `Immutable.Map {}`;

exports[`posts reducer (2) should create group key according to location. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (3) should create group key according to location. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (4) should  not create group for non-filter path. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (5) start fetching "trending". action:POSTS_FETCH_BEGIN 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": true,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (6) start fetching "trending". should keep old entries. action:POSTS_FETCH_BEGIN 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "12": Object {
        "body": "bar",
      },
      "34": Object {
        "body": "foo",
      },
    },
    "err": null,
    "loading": true,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (7) fetching "trending" completed. action:POSTS_FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "14": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
      "12": Object {
        "author": "talhasch",
        "id": 12,
        "permlink": "lorem ipsum",
        "votes": 12,
      },
      "16": Object {
        "author": "chrisbolten",
        "id": 16,
        "permlink": "sit amet",
        "votes": 3,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (8) should not add exiting item to entries. action:POSTS_FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "12": Object {
        "author": "good-karma",
        "id": 12,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
}
`;

exports[`posts reducer (9) simulate load more. action:POSTS_FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "14": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
      "12": Object {
        "author": "talhasch",
        "id": 12,
        "permlink": "lorem ipsum",
        "votes": 12,
      },
      "16": Object {
        "author": "chrisbolten",
        "id": 16,
        "permlink": "sit amet",
        "votes": 3,
      },
      "3": Object {
        "author": "dunsky",
        "id": 3,
        "permlink": "lorem",
        "votes": 31,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (10) error occurred while fetching "trending". action:POSTS_FETCH_ERROR 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "14": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
    },
    "err": "an error has occurred",
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (11) invalidate "trending". action:POSTS_INVALIDATE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`posts reducer (12) should reset error. action:POSTS_INVALIDATE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": true,
    "hasMore": true,
  },
}
`;
